reservation_units/querysets.py

```diff
@@ -0,0 +1,418 @@
+from elasticsearch_django.models import SearchResultsQuerySet
+
+from applications.choices import ApplicationRoundStatusChoice
+from applications.models import ApplicationRound
+from opening_hours.models import ReservableTimeSpan
+from opening_hours.utils.reservable_time_span_client import TimeSpanElement, override_reservable_with_closed_time_spans
+
+if TYPE_CHECKING:
+    from opening_hours.models import ReservationUnit
+
+DEFAULT_TIMEZONE = get_default_timezone()
+
+INTERVAL_TO_MINUTES: dict[str, int] = {
+    "interval_15_mins": 15,
+    "interval_30_mins": 30,
+    "interval_60_mins": 60,
+    "interval_90_mins": 90,
+}
+
+
+def _get_hard_closed_time_spans_for_reservation_unit(reservation_unit: "ReservationUnit") -> list[TimeSpanElement]:
+    reservation_unit_closed_time_spans: list[TimeSpanElement] = []
+
+    if reservation_unit.reservation_begins:
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=datetime.min.replace(tzinfo=DEFAULT_TIMEZONE),
+                end_datetime=reservation_unit.reservation_begins,
+                is_reservable=False,
+            )
+        )
+    if reservation_unit.reservation_ends:
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=reservation_unit.reservation_ends,
+                end_datetime=datetime.max.replace(tzinfo=DEFAULT_TIMEZONE),
+                is_reservable=False,
+            )
+        )
+
+    if reservation_unit.publish_ends:
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=reservation_unit.publish_ends,
+                end_datetime=datetime.max.replace(tzinfo=DEFAULT_TIMEZONE),
+                is_reservable=False,
+            )
+        )
+
+    for application_round in reservation_unit.application_rounds.all():
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=datetime.combine(
+                    application_round.reservation_period_begin, time.min, tzinfo=DEFAULT_TIMEZONE
+                ),
+                end_datetime=datetime.combine(
+                    application_round.reservation_period_end, time.min, tzinfo=DEFAULT_TIMEZONE
+                )
+                + timedelta(days=1),
+                is_reservable=False,
+            )
+        )
+
+    return reservation_unit_closed_time_spans
+
+
+def _get_soft_closed_time_spans_for_reservation_unit(reservation_unit: "ReservationUnit") -> list[TimeSpanElement]:
+    now = timezone.localtime()
+    reservation_unit_closed_time_spans: list[TimeSpanElement] = []
+
+    if reservation_unit.reservations_min_days_before:
+        reservation_unit_closed_time_spans.append(
+            TimeSpanElement(
+                start_datetime=datetime.min.replace(tzinfo=DEFAULT_TIMEZONE),
+                end_datetime=now + timedelta(days=reservation_unit.reservations_min_days_before),
```

Comment:
There is the quirk that (compared to _max_days_) `reservations_min_days_before` is interpreted (intentionally) so that if set, the resulting time truncated to the start (midnight) of the calculated date; i.e., if the time of clock is let's say 15:15 then:
- current date = 2023-05-10 and min_days = 0 => reservation can be made 2023-05-10 @ 15:15
- current date = 2023-05-10 and min_days = 1 => reservation can be made 2023-05-11 @ 00:00
- current date = 2023-05-10 and min_days = 2 => reservation can be made 2023-05-12 @ 00:00
- etc.; cf. https://github.com/City-of-Helsinki/tilavarauspalvelu-core/blob/d795967c6fadff4e6d7a2cc1c00889c32f5a17e4/api/graphql/types/reservations/serializers/mixins.py#L369